src/gh_pr_upsert/git.py

```diff
@@ -0,0 +1,107 @@
+"""Helper functions for working with Git and GitHub."""
+from functools import lru_cache
+from subprocess import CalledProcessError
+
+from gh_pr_upsert.run import run
+
+
+def authenticated_username():
+    """Return the user's authenticated username from GitHub CLI."""
+    return run(["gh", "api", "/user"], json=True)["login"]
+
+
+def pr_diff():
+    """Return the diff of the open PR for the current branch."""
+    return run(["gh", "pr", "diff"])
+
+
+@lru_cache(maxsize=1)
+def pr():  # pylint:disable=invalid-name
+    """Return a dict of info about the open PR for the current branch.
+
+    Return `None` if there's no open PR for the current branch.
+    """
+    try:
+        response = run(["gh", "pr", "view", "--json", "state,commits"], json=True)
+    except CalledProcessError:
+        return None
+
+    if response["state"] != "OPEN":
+        return None
+
+    return response
+
+
+@lru_cache(maxsize=1)
+def github_repo():
+    """Return a dict of info about the GitHub repo."""
+    return run(
+        ["gh", "repo", "view", "--json", "url,sshUrl,defaultBranchRef"], json=True
+    )
+
+
+def default_branch():
+    """Return the name of the GitHub repo's default branch."""
+    return github_repo()["defaultBranchRef"]["name"]
+
+
+def push_urls():
+    """Return a list of the GitHub repo's push URLs."""
+    return [github_repo()["url"] + ".git", github_repo()["sshUrl"]]
+
+
+@lru_cache(maxsize=1)
+def current_branch():
+    """Return the name of the current git branch."""
+    return run(["git", "symbolic-ref", "--quiet", "--short", "HEAD"])
+
+
+def committed_changes():
+    """Return the diff of the current local branch to the GitHub default branch."""
+    return run(["git", "diff", f"{remote()}/{default_branch()}..."])
+
+
+def pr_committers():
+    """Return the GitHub usernames of all committers to the current branch's PR."""
+    committers = set()
+    for commit_ in pr()["commits"]:
+        for author in commit_["authors"]:
+            committers.add(author["login"])
+    return committers
+
+
+def remotes():
+    """Return a list of the local git repo's remotes."""
+    return run(["git", "remote"]).split()
+
+
+def push_url(remote_):
+    """Return the push URL of the given remote."""
+    return run(["git", "remote", "get-url", "--push", remote_])
+
+
+def remote():
+    """
+    Return the local git remote name for the GitHub repo, e.g. "origin".
+
+    Return the name of the local git remote that corresponds to the GitHub repo
+    that GitHub CLI chooses for this git repo.
+    """
+    for remote_ in remotes():
+        if push_url(remote_) in push_urls():
+            return remote_
+
+    raise ValueError("Couldn't find a local git remote that matches the GitHub repo")
```

Comment:
Which remote does GitHub CLI create a pull request on when gh-pr-upsert runs `gh pr create`?

In many cases the local repo will have only one remote, named `origin`, whose URL is that of the GitHub repo that was cloned. But other cases are common. For example you might have forked a repo to your personal account and cloned that. In that case your `origin` remote might be your fork and you may have the original repo as a second remote named `upstream`. Or some other setup.

I have no idea how GitHub CLI chooses which repo to create PRs on in cases like this.

The solution I've implemented is to run `gh repo view` which returns some information about "the repository for the current directory" (according to [GitHub CLI's docs](https://cli.github.com/manual/gh_repo_view)). This is presumably the repo that `gh pr create` creates PR in. And then I look through all the local git repo's remotes and find the one that matches this GitHub repo.

This is awkward AF but I couldn't see another way.